File: 2966-DivideArrayIntoArraysWithMaxDifference/DivideArrayIntoArraysWithMaxDifference.go
package dividearrayintoarrayswithmaxdifference

import (
	"sort"
)

func divideArray(nums []int, k int) [][]int {
	sort.Ints(nums)
	n := len(nums)
	grps := n / 3
	ans := make([][]int, grps)

	i := n - 1
	grps--
	for i > 0 {
		for j := 0; j < 3; j++ {
			if nums[i]-nums[i-2] > k {
				return [][]int{}
			}
			ans[grps] = make([]int, 3)
			ans[grps][2] = nums[i]
			ans[grps][1] = nums[i-1]
			ans[grps][0] = nums[i-2]
		}
		i -= 3
		grps--
	}

	return ans
}

File: 2966-DivideArrayIntoArraysWithMaxDifference/DivideArrayIntoArraysWithMaxDifference.java
import java.util.*;

class DivideArrayIntoArraysWithMaxDifference {
    
    class Solution {
        public int[][] divideArray(int[] nums, int k) {
            Arrays.sort(nums);
            int n = nums.length;
            int grp = n / 3;
            int[][] ans = new int[grp][3];
            
            int i = nums.length - 1;
            
            while(i > 0){
                if(nums[i] - nums[i-2] > k) return new int[][]{};
                
                for(int j = 0; j < 3; j++){
                    ans[grp - 1][2 - j] = nums[i - j];
                }
                
                i -= 3;
                grp--;
            }
            
            return ans;
        }
    }
}
